### ISSUE TITLE
[F4 Research] Local decision model comparator — Laya / Kev (secondary, non-blocking)

### ISSUE THREAD

**Issue by James3014** [James3014/nexus-opencli-reviewer#38]
# F4 Local Decision Model Comparator Track — durable continuation record

## Purpose

建立一條 **SECONDARY / non-blocking** 的 F4 local decision model comparator research track，用相同 F4 canonical semantic evidence 比較 hosted Jev、Laya、Kev。

真正要回答的問題不是哪個 Jev-like model 最熱門，而是：

1. Hosted Jev 是否仍是 Nexus F4 narrow role 的最佳 engineering choice？
2. Laya 是否能在 **same required safety** 下提供 materially lower latency、zero external data egress 與可接受 maintenance cost？
3. Kev 是否能在 **same required safety** 下提供更好的 OOD robustness 與 local deployment？
4. 如果 local models 都無法達標，是否反過來支持 hosted Jev 的額外 capability 值得 provider dependency？

This Issue is a durable research/continuation home. It does **not** grant merge, production, routing, model-promotion, or certification authority.

---

## Current evidence watermark — 2026-09-21

Repository:

`James3014/nexus-opencli-reviewer`

Observed default branch:

- main: `49616ad7d52f44d1054c1ca7e1b4d66305af867a`
- tree: `b364d2f80832e624d84baa07579b328726194480`

Relevant active PR:

- PR #37 — `feat(f4): Context Economics Wave 1 harness and simulation matrix`
- state: OPEN / DRAFT / NOT MERGED
- current head: `5cbc207dc54196d2bd46a5949b85ac0f55a5b429`
- base: `49616ad7d52f44d1054c1ca7e1b4d66305af867a`

At PR #37 head, source physically contains the useful F4 separation:

- `ProviderVisibleRiskStateV1`
- `F4QuestionContractV1`
- provider-neutral `RiskModelAdapter`
- raw `RiskModelResult`
- separate `FamilyCalibrationPolicy`
- hosted/local transport surfaces

This is evidence that the comparator can likely reuse the existing adapter boundary.

**Important:** PR #37 is not merged. Do not treat its branch source as accepted main/runtime authority.

---

## Separate lineage: completed Context Economics / Context Relevance live diagnostic

This is **not** F4 Risk Sensor certification evidence and must remain a separate lineage.

Completed exact live diagnostic:

- Candidate: `5cbc207dc54196d2bd46a5949b85ac0f55a5b429`
- authorization hash: `27089abb52469b6a0e2896558f54f297a298e1380f1b4bda2553b4eea4293ce7`
- exact samples: 25
- provider attempts: 25
- `OBSERVED_OK`: 25/25
- `OUTCOME_UNKNOWN`: 0
- retry: 0
- private result SHA-256: `d4bf234fd079c6b1d5091e334eb889ee027d44a95ae05d34b084000fe66c2fac`
- measured Context Relevance ranking signal:
  - Spearman vs fixture relevance: ~0.932
  - Pearson: ~0.973
  - high-vs-low AUC: 1.000
  - materially different pair ordering: 200/200

Supported claim ceiling:

`REAL_JEV_RANKING_SIGNAL_MEASURED`

Not supported by this experiment:

- F4 Risk Sensor certification
- calibrated probability semantics
- production threshold
- automatic context deletion
- general semantic reflex
- routing
- production readiness

Do not reuse/relabel these 25 Context Relevance cases as F4 Risk Sensor calibration/certification data.

---

# PRIMARY — unchanged

The primary line remains:

```text
F4 Risk Sensor
→ Jev Hosted Adapter
→ privacy-safe calibration
→ prospective certification
```

The local comparator track MUST NOT delay or block this line.

Do not change the primary F4 question, evidence projection, shadow seam, calibration lineage, prospective case selection, or production authority merely because Laya/Kev exist.

---

# SECONDARY — F4 LOCAL DECISION MODEL COMPARATOR TRACK

Candidate models:

### Laya

Research hypothesis:

- local / open source
- encoder-style / non-autoregressive decision model
- very low latency
- generic zero-shot typed-decision ability may be weak
- specialized checkpoint/fine-tuning may account for much of its typed-decision performance
- cardinality increase may degrade performance
- calibration may require domain refit

First question is NOT “is Laya faster than Jev?”

First question:

> Is the frozen F4 `ProviderVisibleRiskStateV1` semantic signal sufficient for a public, unmodified Laya checkpoint?

If critical misses are unacceptable, speed is irrelevant safety evidence.

### Kev

Research hypothesis:

- local
- Qwen backbone + LoRA + pointer/readout head
- prefill-only / no text decoding
- typed Noul / Choice / Score style decisions
- TypeSafe-compatible API concept
- OOD / semantic composition is a primary research interest
- Kev-4B remains research preview and must not be treated as certified

Kev evaluation must include:

- disguised dangerous
- looks-benign-but-dangerous
- unseen combinations
- permission/risk composition
- evidence perturbation

High in-domain score cannot hide OOD degradation.

### Nimble

Status:

`REFERENCE / TRAINING-RECIPE REVIEW`

Do not include Nimble in first comparator execution.

Use it only as a reference for:

- training-data construction
- hard negatives
- candidate scoring
- readout architecture
- calibration techniques

No domain fine-tuning in first comparator phase.

---

# Hard architecture boundary: same evidence universe

All adapters must consume exactly the same canonical F4 semantic evidence:

```text
SafeRiskObservation
        ↓
ProviderVisibleRiskStateV1
        ├── Jev Adapter
        ├── Laya Adapter
        └── Kev Adapter
```

Forbidden:

- giving Laya extra raw text
- giving Kev extra objectives/history
- giving local models paths/diffs/task lineage unavailable to Jev
- using different ground truth
- silently changing F4 wording for one provider
- enriching one model's evidence universe

Provider-specific serialization is allowed.

Provider-specific semantic evidence is not.

Add an evidence-parity check based on the canonical provider-visible payload/hash before comparing model behavior.

---

# Adapter contract

Prefer reusing the existing provider-neutral F4 interface rather than creating a new authority:

```python
RiskModelAdapter.evaluate(
    state: ProviderVisibleRiskStateV1,
    contract: F4QuestionContractV1,
    provider_request_id: str,
) -> RiskModelResult
```

Target adapter shape:

```text
RiskModelAdapter
├── JevHostedRiskModelAdapter
├── LayaRiskModelAdapter
└── KevRiskModelAdapter
```

Generic adapter returns only raw/model facts such as:

- probability / distribution
- provider status
- adapter/model identity
- exact revision
- provider payload hash
- API/service version if available

Do NOT put an F4 policy threshold in the adapter.

Calibration / threshold / abstention remain outside the generic adapter.

Performance telemetry (latency, RAM, cold start, throughput) belongs in the benchmark harness, not the canonical semantic adapter contract unless later evidence proves otherwise.

---

# Phase 1 — Compatibility / Smoke Gate

First local-model execution gate only.

Do NOT use the held-out F4 evaluation set yet.

For Laya and Kev independently record:

A. exact repository revision
B. exact model/checkpoint revision
C. license
D. architecture
E. input/context/cardinality limits
F. output contract
G. Apple Silicon compatibility
H. runtime/backend requirements
I. RAM / model size
J. cold start
K. warm deterministic repeatability
L. offline capability
M. semantic compatibility with the frozen `F4QuestionContractV1`
N. evidence parity with `ProviderVisibleRiskStateV1`

Smoke-gate result:

- `COMPATIBLE_FOR_F4_OFFLINE_COMPARATOR`
- `DEFER`
- `REJECT`

Do not produce a winner in Phase 1.

---

# Phase 2 — Raw F4 signal evaluation

Only models that pass compatibility smoke proceed.

Matched inputs:

```text
JEV_F4
LAYA_F4
KEV_F4
```

First ask whether the raw model has useful F4 risk signal before tuning thresholds.

## HIGH-RISK

- critical misses
- selective risk
- risk UCB
- coverage
- abstention behavior
- disguised-dangerous cases
- benign-looking-dangerous cases
- unseen combinations
- permission/risk composition
- evidence perturbation

## LOW-RISK

- false escalation
- burden reduction

## Robustness

- missing evidence
- contradictory evidence
- reordered/irrelevant evidence
- deterministic repeatability
- cardinality/input representation stress

Do not select a production threshold in this phase.

---

# Phase 3 — Calibration separation

Calibration and evaluation must be physically/logically separate.

Per model:

```text
CALIBRATION
→ freeze model-specific policy
→ SEALED HELD_OUT / PROSPECTIVE
```

Each model may have its own:

- temperature
- threshold
- abstention region

All models must share:

- same ground truth
- same frozen evaluation cases
- same F4 semantics
- same evidence universe

Forbidden:

- tune temperature on held-out then report that held-out as independent
- tune threshold on held-out
- change question wording after seeing results
- change mapping after seeing results
- global threshold across unrelated model families

Existing Jev certification/calibration lineage must remain untouched.

If local comparator requires an abstention policy not expressible by current `FamilyCalibrationPolicy`, first propose a bounded policy-extension change. Do not mutate Jev certification semantics implicitly.

---

# Phase 4 — Sealed offline comparator

Required metrics:

## HIGH-RISK
- critical misses
- selective risk
- risk UCB
- coverage
- abstention

## LOW-RISK
- false escalation
- burden reduction

## CALIBRATION
- Brier
- ECE
- reliability by threshold/family
- missing/contradictory evidence behavior

## SYSTEM
- warm p50
- warm p95
- warm p99
- cold start
- RAM
- CPU/GPU/unified memory
- throughput
- availability
- local/offline capability

## OPERATIONS
- provider dependency
- privacy/data egress
- update drift
- maintenance burden
- failure surface

Final comparison must evaluate:

`quality at required safety + latency + privacy + cost + operational complexity`

Not accuracy alone.

---

# Failure classification before any fine-tuning

If a local model fails, classify why before proposing adaptation:

- `SIGNAL_INSUFFICIENT`
- `MODEL_CAPACITY`
- `CALIBRATION`
- `QUESTION_CONTRACT`
- `OOD_GENERALIZATION`
- `CARDINALITY`
- `INPUT_REPRESENTATION`

Do NOT immediately fine-tune Nexus F4 data.

Only if evidence supports:

1. F4 is a narrow learnable task, and
2. failure is primarily model adaptation rather than semantic/input insufficiency,

may a separate proposal be created for:

`DOMAIN_SPECIALIZED_F4_MODEL`

That is a separate gate and must not be auto-started.

---

# Success claim ceiling

If a local model passes the offline comparator, maximum claim is:

`LOCAL_F4_OPTIONAL_ADAPTER_CANDIDATE`

Never infer:

- `REPLACE_JEV`
- `PRODUCTION_READY`
- `F4_MODEL_CERTIFIED`

A passing model must still undergo an independent passive prospective window before any stronger claim.

Production/certification authority remains unchanged.

---

# Decision tree

```text
PRIMARY:
Jev prospective certification
continues independently.

SECONDARY:
Laya / Kev
↓
compatibility smoke
↓
offline F4 calibration
↓
sealed evaluation

both fail
→ LOCAL_COMPARATOR_DEFER
→ Jev mainline continues

one passes
→ that model gets independent prospective passive shadow

both pass
→ both enter passive prospective comparison
```

No automatic promotion.

---

# General Semantic Reflex remains closed

Existing decision remains:

```text
STOP_SEMANTIC_REFLEX_EXPANSION
GENERAL_SEMANTIC_REFLEX = NOT_SUPPORTED
ADOPTION = DEFER
```

Laya/Kev are NOT a reopen condition.

Only this narrow F4 optional-adapter comparator is allowed.

Do not reopen:

- General Semantic Reflex
- general routing
- semantic permission authority
- model-routing experimentation

---

# Jev ecosystem patterns to review separately

Research-only references:

### jevcal
Inspect:
- threshold lock
- calibration provenance
- drift detection

If stronger than current F4 methodology, propose a bounded delta only.

### jev-ood-calibration
Inspect:
- OOD design
- family-specific calibration
- confident-error behavior

Use as methodology input only.

### huncho
Inspect:
- hysteresis
- journal
- replay discipline

Do not assume it is stronger than current Nexus effect identity/replay rules.

### pi-jev-auto-mode
Inspect:
- deterministic-first + ambiguous semantic judgment pattern

Do not transfer permission authority to a semantic model.

### jev-router
REFERENCE ONLY.

Do not open a Nexus routing experiment from this track.

---

# Physical evidence lineage

Keep local comparator evidence separate from Jev certification evidence.

Suggested layout:

```text
f4-local-comparator/
├── manifest.json
├── evidence-parity.json
├── laya/
│   ├── compatibility/
│   ├── calibration/
│   ├── held-out/
│   └── prospective/
└── kev/
    ├── compatibility/
    ├── calibration/
    ├── held-out/
    └── prospective/
```

Do not move/relabel/reselect existing Jev prospective cases after seeing local-model results.

---

# Execution order

Do not interrupt PRIMARY.

```text
PRIMARY
1. Finish current Jev F4 gate
2. freeze exact Jev evidence/revision
3. privacy-safe calibration
4. prospective certification

SECONDARY
C0. create isolated local-comparator branch/worktree
C1. Laya revision/license/runtime compatibility smoke
C2. Kev revision/license/runtime compatibility smoke
C3. semantic-contract fidelity + evidence-parity audit
C4. raw F4 signal evaluation
C5. calibration split and per-model freeze
C6. sealed held-out evaluation
C7. system benchmark
C8. Jev vs Laya vs Kev ROLE FIT MATRIX
C9. verdict: BOUNDED_PILOT / DEFER / REJECT
C10. only if justified: passive prospective local shadow proposal
```

Do not simultaneously make broad Hosted Adapter and Local Adapter changes.

---

# Required first-round report

A. exact model revisions
B. licenses
C. runtime requirements
D. adapter contract
E. evidence parity audit
F. calibration separation
G. high-risk metrics
H. low-risk utility
I. OOD results
J. latency / RAM / cold start
K. privacy / operations comparison
L. Jev vs Laya vs Kev ROLE FIT MATRIX
M. exact verdict for each:
   - `BOUNDED_PILOT`
   - `DEFER`
   - `REJECT`
N. one next gate

Hard stop:

```text
NO PUSH
NO MERGE
NO PRODUCTION AUTHORITY CHANGE
```

---

# Current frontier / continuation note

This Issue is intentionally **SECONDARY** and **non-blocking**.

Do not begin comparator mutation merely because this Issue exists.

Before C0/C1 implementation:

1. rebind current repository/main;
2. identify the exact frozen F4 Risk Sensor revision/evidence lineage;
3. confirm PRIMARY Jev work is not being delayed;
4. create an isolated comparator worktree/branch;
5. preserve evidence parity and no-production-authority boundaries.

Until that start gate is satisfied, research can be treated as `ANALYSIS_ONLY`.

## Maximum claim

This Issue records an Owner-approved research direction and continuation contract only.

It does not prove:
- Laya compatibility
- Kev compatibility
- local-model safety
- local-model certification
- Jev replacement
- merge readiness
- production readiness
